src/libraries/System.Collections/src/System/Collections/Generic/SortedDictionary.cs

```diff
@@ -35,11 +35,21 @@ public SortedDictionary(IDictionary<TKey, TValue> dictionary, IComparer<TKey>? c
                 throw new ArgumentNullException(nameof(dictionary));
             }
 
-            _set = new TreeSet<KeyValuePair<TKey, TValue>>(new KeyValuePairComparer(comparer));
+            var keyValuePairComparer = new KeyValuePairComparer(comparer);
 
-            foreach (KeyValuePair<TKey, TValue> pair in dictionary)
+            if (dictionary is SortedDictionary<TKey, TValue> sortedDictionary &&
+                sortedDictionary._set.Comparer.Equals(keyValuePairComparer))
```

Comment:
Rather than changing the equality for KeyValuePairComparer consider the following change:

```suggestion
                sortedDictionary._set.Comparer is KeyValuePairComparer kv &&
                kv.KeyComparer.Equals(keyValuePairComparer.KeyComparer)
```